File: app/src/main/java/com/jeanwest/reader/findingResultSubActivity.java
package com.jeanwest.reader;

import android.annotation.SuppressLint;
import android.media.AudioManager;
import android.media.ToneGenerator;
import android.os.Bundle;
import android.os.Handler;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.CheckBox;
import android.widget.SeekBar;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import com.rscja.deviceapi.RFIDWithUHFUART;
import com.rscja.deviceapi.entity.UHFTAGInfo;
import com.rscja.deviceapi.exception.ConfigurationException;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.Map;

public class findingResultSubActivity extends AppCompatActivity {

    public RFIDWithUHFUART RF;
    public static Map<String, Integer> EPCTableFinding = new HashMap<>();
    ToneGenerator beep = new ToneGenerator(AudioManager.STREAM_MUSIC, 100);
    private int findingPower = 30;
    TextView status;
    TextView powerText;
    SeekBar powerSeekBar;
    TextView numberOfFoundText;
    TextView stuffSpec;
    private long stuffCode;
    WebView picture;
    CheckBox option;

    public static String stuffPrimaryCode;
    public static String stuffRFIDCode;
    JSONObject stuff;
    Map<String, Integer> EPCTableFindingMatched = new HashMap<>();
    WebSettings setting;

    boolean isChecked = false;
    boolean findingInProgress = false;
    boolean readEnable = false;

    Handler databaseBackgroundTaskHandler = new Handler();

    Runnable databaseBackgroundTask = new Runnable() {
        @SuppressLint("SetTextI18n")
        @Override
        public void run() {

            if(findingInProgress) {

                UHFTAGInfo uhftagInfo;

                while (true) {

                    uhftagInfo = RF.readTagFromBuffer();

                    if (uhftagInfo != null) {
                        EPCTableFinding.put(uhftagInfo.getEPC(), 1);
                    } else {
                        break;
                    }
                }

                if(EPCTableFinding.size() > 0) {

                    flag = false;

                    for (Map.Entry<String, Integer> EPC : EPCTableFinding.entrySet()) {

                        EPCHexString = EPC.getKey();
                        EPCInt1 = Long.parseLong(EPCHexString.substring(0, 8), 16);
                        EPCInt2 = Long.parseLong(EPCHexString.substring(8, 16), 16);
                        EPCInt3 = Long.parseLong(EPCHexString.substring(16, 24), 16);

                        EPCBinaryString1 = Long.toBinaryString(EPCInt1);
                        EPCBinaryString1 = String.format("%32s", EPCBinaryString1).replaceAll(" ", "0");
                        EPCBinaryString2 = Long.toBinaryString(EPCInt2);
                        EPCBinaryString2 = String.format("%32s", EPCBinaryString2).replaceAll(" ", "0");
                        EPCBinaryString3 = Long.toBinaryString(EPCInt3);
                        EPCBinaryString3 = String.format("%32s", EPCBinaryString3).replaceAll(" ", "0");

                        EPCBinaryString = EPCBinaryString1 + EPCBinaryString2 + EPCBinaryString3;

                        header = Integer.parseInt(EPCBinaryString.substring(0, 8), 2);
                        companyNumber = Integer.parseInt(EPCBinaryString.substring(14, 26), 2);
                        itemNumber = Long.parseLong(EPCBinaryString.substring(26, 58), 2);

                        if (companyNumber == 100) {
                            stuffCode = Long.parseLong(stuffPrimaryCode);
                        } else if (companyNumber == 101) {
                            stuffCode = Long.parseLong(stuffRFIDCode);
                        }
                        if (header == 48 && itemNumber == stuffCode) {
                            EPCTableFindingMatched.put(EPCHexString, 1);
                            flag = true;
                        }
                    }

                    numberOfFoundText.setText(String.valueOf(EPCTableFindingMatched.size()));
                    status.setText("");
                    status.setText("در حال جست و جو ...");

                    for(Map.Entry<String, Integer> valid : EPCTableFindingMatched.entrySet()) {
                        status.setText(status.getText() + "\n" + valid.getKey());
                    }

                    if (flag) {
                        beep.startTone(ToneGenerator.TONE_CDMA_PIP, 500);
                    }

                    EPCTableFinding.clear();
                    databaseBackgroundTaskHandler.postDelayed(databaseBackgroundTask, 1000);
                }
                else {
                    databaseBackgroundTaskHandler.postDelayed(databaseBackgroundTask, 1000);
                }
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_finding_result_sub);
        status = findViewById(R.id.section_label0);
        stuffSpec = findViewById(R.id.result0);
        picture = findViewById(R.id.pictureView0);
        powerText = findViewById(R.id.findingPowerTextView0);
        powerSeekBar = findViewById(R.id.findingPowerSeekBar0);
        numberOfFoundText = findViewById(R.id.numberOfFoundTextView0);
        option = findViewById(R.id.checkBox0);

        powerSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @SuppressLint("SetTextI18n")
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

                if (!readEnable) {

                    switch(progress) {
                        case 0:
                            findingPower = 5;
                            break;
                        case 1:
                            findingPower = 10;
                            break;
                        case 2:
                            findingPower = 15;
                            break;
                        case 3:
                            findingPower = 20;
                            break;
                        case 4:
                            findingPower = 30;
                            break;
                    }

                    powerText.setText("قدرت سیگنال(" + findingPower + ")");
                } else {
                    RF.stopInventory();
                    switch(progress) {
                        case 0:
                            findingPower = 5;
                            break;
                        case 1:
                            findingPower = 10;
                            break;
                        case 2:
                            findingPower = 15;
                            break;
                        case 3:
                            findingPower = 20;
                            break;
                        case 4:
                            findingPower = 30;
                            break;
                    }
                    powerText.setText("قدرت سیگنال(" + findingPower + ")");
                    while (!RF.setPower(findingPower)) {}
                    RF.startInventoryTag(0, 0, 0);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        try {
            RF = RFIDWithUHFUART.getInstance();
        } catch (ConfigurationException e) {
            e.printStackTrace();
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    protected void onResume() {
        super.onResume();

        while (!RF.setPower(findingPower)) {
        }

        try {

            stuff = findingActivity.API.similar.getJSONObject(findingActivity.index);
            stuffSpec.setText(stuff.getString("productName") + "\n" +
                    "کد محصول: " + stuff.getString("K_Bar_Code") + "\n" +
                    "بارکد: " + stuff.getString("KBarCode") + "\n" +
                    "قیمت مصرف کننده: " + stuff.getString("OrigPrice") + "\n" +
                    "قیمت فروش: " + stuff.getString("SalePrice") + "\n" +
                    "موجودی فروشگاه: " + stuff.getString("dbCountStore") + "\n" +
                    "موجودی انبار: " + stuff.getString("dbCountDepo"));

            picture.loadUrl(stuff.getString("ImgUrl"));
            stuffPrimaryCode = stuff.getString("BarcodeMain_ID");
            stuffRFIDCode = stuff.getString("RFID");
        } catch (JSONException e) {
            e.printStackTrace();
        }

        setting = picture.getSettings();
        setting.setUseWideViewPort(true);
        setting.setLoadWithOverviewMode(true);
        picture.setFocusable(false);

        reading.databaseInProgress = false;

        switch(findingPower) {
            case 5:
                powerSeekBar.setProgress(0);
                break;
            case 10:
                powerSeekBar.setProgress(1);
                break;
            case 15:
                powerSeekBar.setProgress(2);
                break;
            case 20:
                powerSeekBar.setProgress(3);
                break;
            case 30:
                powerSeekBar.setProgress(4);
                break;
        }
        powerText.setText("قدرت سیگنال(" + findingPower + ")");

    }

    int header;
    int companyNumber;
    long itemNumber;
    boolean flag;
    long EPCInt1;
    long EPCInt2;
    long EPCInt3;
    String EPCHexString;
    String EPCBinaryString;
    String EPCBinaryString1;
    String EPCBinaryString2;
    String EPCBinaryString3;

    @SuppressLint("SetTextI18n")
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {

        if (keyCode == 280 || keyCode == 139 || keyCode == 293) {

            if (event.getRepeatCount() == 0) {

                status.setText("");

                if (!readEnable) {

                    EPCTableFinding.clear();
                    if(!isChecked) {
                        EPCTableFindingMatched.clear();
                    }

                    while (!RF.setPower(findingPower)) {
                    }
                    readEnable = true;
                    RF.startInventoryTag(0,0 ,0);

                    status.setText("در حال جست و جو ...");
                    findingInProgress = true;
                    databaseBackgroundTaskHandler.postDelayed(databaseBackgroundTask, 1000);
                }
                else {

                    databaseBackgroundTaskHandler.removeCallbacks(databaseBackgroundTask);

                    readEnable = false;
                    RF.stopInventory();

                    findingInProgress = false;

                    flag = false;

                    for (Map.Entry<String, Integer> EPC : EPCTableFinding.entrySet()) {

                        EPCHexString = EPC.getKey();
                        EPCInt1 = Long.parseLong(EPCHexString.substring(0, 8), 16);
                        EPCInt2 = Long.parseLong(EPCHexString.substring(8, 16), 16);
                        EPCInt3 = Long.parseLong(EPCHexString.substring(16, 24), 16);

                        EPCBinaryString1 = Long.toBinaryString(EPCInt1);
                        EPCBinaryString1 = String.format("%32s", EPCBinaryString1).replaceAll(" ", "0");
                        EPCBinaryString2 = Long.toBinaryString(EPCInt2);
                        EPCBinaryString2 = String.format("%32s", EPCBinaryString2).replaceAll(" ", "0");
                        EPCBinaryString3 = Long.toBinaryString(EPCInt3);
                        EPCBinaryString3 = String.format("%32s", EPCBinaryString3).replaceAll(" ", "0");

                        EPCBinaryString = EPCBinaryString1 + EPCBinaryString2 + EPCBinaryString3;

                        header = Integer.parseInt(EPCBinaryString.substring(0, 8), 2);
                        companyNumber = Integer.parseInt(EPCBinaryString.substring(14, 26), 2);
                        itemNumber = Long.parseLong(EPCBinaryString.substring(26, 58), 2);

                        if (companyNumber == 100) {
                            stuffCode = Long.parseLong(stuffPrimaryCode);
                        } else if (companyNumber == 101) {
                            stuffCode = Long.parseLong(stuffRFIDCode);
                        }
                        if (header == 48 && itemNumber == stuffCode) {

                            EPCTableFindingMatched.put(EPCHexString, 1);
                            flag = true;
                        }
                    }
                    numberOfFoundText.setText(String.valueOf(EPCTableFindingMatched.size()));

                    status.setText("");
                    for(Map.Entry<String, Integer> valid : EPCTableFindingMatched.entrySet()) {
                        status.setText(status.getText() + "\n" + valid.getKey());
                    }

                    if (flag) {
                        beep.startTone(ToneGenerator.TONE_CDMA_PIP, 500);
                    }

                    EPCTableFinding.clear();
                }
            }

        } else if (keyCode == 4) {

            if(readEnable) {
                RF.stopInventory();
                readEnable = false;
            }

            findingInProgress = false;
            databaseBackgroundTaskHandler.removeCallbacks(databaseBackgroundTask);

            finish();
        }
        return true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        if(readEnable) {
            RF.stopInventory();
            readEnable = false;
        }
    }

    public void clearEPCs(View view) {

        EPCTableFinding.clear();
        EPCTableFindingMatched.clear();
        status.setText("");
        numberOfFoundText.setText("0");
    }

    public void optionChange(View view) {

        isChecked = option.isChecked();
    }
}

File: app/src/main/java/com/jeanwest/reader/FindingResultSubActivity.kt
package com.jeanwest.reader

import android.annotation.SuppressLint
import android.media.AudioManager
import android.media.ToneGenerator
import android.os.Bundle
import android.os.Handler
import android.view.KeyEvent
import android.view.View
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.CheckBox
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.rscja.deviceapi.RFIDWithUHFUART
import com.rscja.deviceapi.entity.UHFTAGInfo
import com.rscja.deviceapi.exception.ConfigurationException
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class FindingResultSubActivity : AppCompatActivity() {

    lateinit var rf: RFIDWithUHFUART
    var beep = ToneGenerator(AudioManager.STREAM_MUSIC, 100)
    private var findingPower = 30
    lateinit var status: TextView
    lateinit var powerText: TextView
    lateinit var powerSeekBar: SeekBar
    lateinit var numberOfFoundText: TextView
    lateinit var stuffSpec: TextView
    private var stuffCode: Long = 0
    lateinit var picture: WebView
    lateinit var option: CheckBox
    lateinit var stuff: JSONObject
    var epcTableFindingMatched: MutableMap<String?, Int> = HashMap()
    lateinit var setting: WebSettings
    var isChecked = false
    var findingInProgress = false
    var readEnable = false

    var header = 0
    var companyNumber = 0
    var itemNumber: Long = 0
    var flag = false
    var EPCInt1: Long = 0
    var EPCInt2: Long = 0
    var EPCInt3: Long = 0
    lateinit var EPCHexString: String
    lateinit var EPCBinaryString: String
    lateinit var EPCBinaryString1: String
    lateinit var EPCBinaryString2: String
    lateinit var EPCBinaryString3: String
    
    var databaseBackgroundTaskHandler = Handler()
    
    var databaseBackgroundTask: Runnable = object : Runnable {
        
        @SuppressLint("SetTextI18n")
        override fun run() {
            
            if (findingInProgress) {
                
                var uhfTagInfo: UHFTAGInfo?
                
                while (true) {
                    
                    uhfTagInfo = rf.readTagFromBuffer()
                    if (uhfTagInfo != null) {
                        EPCTableFinding[uhfTagInfo.epc] = 1
                    } else {
                        break
                    }
                }
                if (EPCTableFinding.isNotEmpty()) {
                    flag = false
                    for ((key) in EPCTableFinding) {
                        EPCHexString = key
                        EPCInt1 = EPCHexString.substring(0, 8).toLong(16)
                        EPCInt2 = EPCHexString.substring(8, 16).toLong(16)
                        EPCInt3 = EPCHexString.substring(16, 24).toLong(16)
                        EPCBinaryString1 = java.lang.Long.toBinaryString(EPCInt1)
                        EPCBinaryString1 = String.format("%32s", EPCBinaryString1).replace(" ".toRegex(), "0")
                        EPCBinaryString2 = java.lang.Long.toBinaryString(EPCInt2)
                        EPCBinaryString2 = String.format("%32s", EPCBinaryString2).replace(" ".toRegex(), "0")
                        EPCBinaryString3 = java.lang.Long.toBinaryString(EPCInt3)
                        EPCBinaryString3 = String.format("%32s", EPCBinaryString3).replace(" ".toRegex(), "0")
                        EPCBinaryString = EPCBinaryString1 + EPCBinaryString2 + EPCBinaryString3
                        header = EPCBinaryString.substring(0, 8).toInt(2)
                        companyNumber = EPCBinaryString.substring(14, 26).toInt(2)
                        itemNumber = EPCBinaryString.substring(26, 58).toLong(2)
                        if (companyNumber == 100) {
                            stuffCode = stuffPrimaryCode.toLong()
                        } else if (companyNumber == 101) {
                            stuffCode = stuffRFIDCode.toLong()
                        }
                        if (header == 48 && itemNumber == stuffCode) {
                            epcTableFindingMatched[EPCHexString] = 1
                            flag = true
                        }
                    }
                    numberOfFoundText.text = epcTableFindingMatched.size.toString()

                    status.text = "در حال جست و جو ..."
                    for ((key) in epcTableFindingMatched) {
                        status.text = "${status.text}" + "\n$key"
                    }
                    if (flag) {
                        beep.startTone(ToneGenerator.TONE_CDMA_PIP, 500)
                    }
                    EPCTableFinding.clear()
                    databaseBackgroundTaskHandler.postDelayed(this, 1000)
                } else {
                    databaseBackgroundTaskHandler.postDelayed(this, 1000)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_finding_result_sub)
        status = findViewById(R.id.section_label0)
        stuffSpec = findViewById(R.id.result0)
        picture = findViewById(R.id.pictureView0)
        powerText = findViewById(R.id.findingPowerTextView0)
        powerSeekBar = findViewById(R.id.findingPowerSeekBar0)
        numberOfFoundText = findViewById(R.id.numberOfFoundTextView0)
        option = findViewById(R.id.checkBox0)
        powerSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            @SuppressLint("SetTextI18n")
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (!readEnable) {
                    when (progress) {
                        0 -> findingPower = 5
                        1 -> findingPower = 10
                        2 -> findingPower = 15
                        3 -> findingPower = 20
                        4 -> findingPower = 30
                    }
                    powerText.setText("قدرت سیگنال($findingPower)")
                } else {
                    rf.stopInventory()
                    when (progress) {
                        0 -> findingPower = 5
                        1 -> findingPower = 10
                        2 -> findingPower = 15
                        3 -> findingPower = 20
                        4 -> findingPower = 30
                    }
                    powerText.setText("قدرت سیگنال($findingPower)")
                    while (!rf.setPower(findingPower)) {
                    }
                    rf.startInventoryTag(0, 0, 0)
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        try {
            rf = RFIDWithUHFUART.getInstance()
        } catch (e: ConfigurationException) {
            e.printStackTrace()
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onResume() {
        super.onResume()
        while (!rf.setPower(findingPower)) {
        }
        try {
            stuff = FindingProductActivity.API.similar.getJSONObject(FindingProductActivity.index)
            stuffSpec.text = """
                ${stuff.getString("productName")}
                کد محصول: ${stuff.getString("K_Bar_Code")}
                بارکد: ${stuff.getString("KBarCode")}
                قیمت مصرف کننده: ${stuff.getString("OrigPrice")}
                قیمت فروش: ${stuff.getString("SalePrice")}
                موجودی فروشگاه: ${stuff.getString("dbCountStore")}
                موجودی انبار: ${stuff.getString("dbCountDepo")}
                """.trimIndent()
            picture.loadUrl(stuff.getString("ImgUrl"))
            stuffPrimaryCode = stuff.getString("BarcodeMain_ID")
            stuffRFIDCode = stuff.getString("RFID")
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        setting = picture.settings
        setting.useWideViewPort = true
        setting.loadWithOverviewMode = true
        picture.isFocusable = false
        reading.databaseInProgress = false
        when (findingPower) {
            5 -> powerSeekBar.progress = 0
            10 -> powerSeekBar.progress = 1
            15 -> powerSeekBar.progress = 2
            20 -> powerSeekBar.progress = 3
            30 -> powerSeekBar.progress = 4
        }
        powerText.text = "قدرت سیگنال($findingPower)"
    }
    
    @SuppressLint("SetTextI18n")
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == 280 || keyCode == 139 || keyCode == 293) {
            if (event.repeatCount == 0) {
                status.text = ""
                if (!readEnable) {
                    EPCTableFinding.clear()
                    if (!isChecked) {
                        epcTableFindingMatched.clear()
                    }
                    while (!rf.setPower(findingPower)) {
                    }
                    readEnable = true
                    rf.startInventoryTag(0, 0, 0)
                    status.text = "در حال جست و جو ..."
                    findingInProgress = true
                    databaseBackgroundTaskHandler.postDelayed(databaseBackgroundTask, 1000)
                } else {
                    databaseBackgroundTaskHandler.removeCallbacks(databaseBackgroundTask)
                    readEnable = false
                    rf.stopInventory()
                    findingInProgress = false
                    flag = false
                    for ((key) in EPCTableFinding) {
                        EPCHexString = key
                        EPCInt1 = EPCHexString.substring(0, 8).toLong(16)
                        EPCInt2 = EPCHexString.substring(8, 16).toLong(16)
                        EPCInt3 = EPCHexString.substring(16, 24).toLong(16)
                        EPCBinaryString1 = java.lang.Long.toBinaryString(EPCInt1)
                        EPCBinaryString1 = String.format("%32s", EPCBinaryString1).replace(" ".toRegex(), "0")
                        EPCBinaryString2 = java.lang.Long.toBinaryString(EPCInt2)
                        EPCBinaryString2 = String.format("%32s", EPCBinaryString2).replace(" ".toRegex(), "0")
                        EPCBinaryString3 = java.lang.Long.toBinaryString(EPCInt3)
                        EPCBinaryString3 = String.format("%32s", EPCBinaryString3).replace(" ".toRegex(), "0")
                        EPCBinaryString = EPCBinaryString1 + EPCBinaryString2 + EPCBinaryString3
                        header = EPCBinaryString.substring(0, 8).toInt(2)
                        companyNumber = EPCBinaryString.substring(14, 26).toInt(2)
                        itemNumber = EPCBinaryString.substring(26, 58).toLong(2)
                        if (companyNumber == 100) {
                            stuffCode = stuffPrimaryCode.toLong()
                        } else if (companyNumber == 101) {
                            stuffCode = stuffRFIDCode.toLong()
                        }
                        if (header == 48 && itemNumber == stuffCode) {
                            epcTableFindingMatched[EPCHexString] = 1
                            flag = true
                        }
                    }
                    numberOfFoundText.text = epcTableFindingMatched.size.toString()
                    status.text = ""
                    for ((key) in epcTableFindingMatched) {
                        status.text = "${status.text}" + "\n$key"
                    }
                    if (flag) {
                        beep.startTone(ToneGenerator.TONE_CDMA_PIP, 500)
                    }
                    EPCTableFinding.clear()
                }
            }
        } else if (keyCode == 4) {
            if (readEnable) {
                rf.stopInventory()
                readEnable = false
            }
            findingInProgress = false
            databaseBackgroundTaskHandler.removeCallbacks(databaseBackgroundTask)
            finish()
        }
        return true
    }

    override fun onPause() {
        super.onPause()
        if (readEnable) {
            rf.stopInventory()
            readEnable = false
        }
    }

    fun clearEPCs(view: View?) {
        EPCTableFinding.clear()
        epcTableFindingMatched.clear()
        status.text = ""
        numberOfFoundText.text = "0"
    }

    fun optionChange(view: View?) {
        isChecked = option.isChecked
    }

    companion object {
        var EPCTableFinding: MutableMap<String, Int> = HashMap()
        lateinit var stuffPrimaryCode: String
        lateinit var stuffRFIDCode: String
    }
}